stdlib/public/core/StringCharacterView.swift

```diff
@@ -188,11 +311,17 @@ extension String: BidirectionalCollection {
   ///
   /// - Parameter i: A valid index of the string. `i` must be less than the
   ///   string's end index.
-  @inlinable @inline(__always)
+  @inlinable @inline(__always) // FIXME(lorentey): Consider removing these. If
+                               // `index(after:)` isn't inlinable, does it
+                               // really matter if this one is? (Potential
+                               // _guts-related optimizations notwithstanding.)
+                               // `subscript` being inlinable forces a bunch of
+                               // new additions to be _aEIC, even though they
+                               // ought to be internal.
   public subscript(i: Index) -> Character {
+    var i = _guts.ensureMatchingEncoding(i)
```

Comment:
Maybe we can coalesce bounds checking with encoding checking.

Reply:
This turned out to be very much possible and I think it made things a lot clearer! 🎉

There was no need for any closure-based approach. 922a477dbcb1b652d561ebb584bb16cbd252133a introduces 2 * 3 new index validation methods that perform all three steps above.